test/Dialect/FIRRTL/lower-layers.mlir

```diff
@@ -471,3 +471,82 @@ firrtl.circuit "ModuleNameConflict" {
 // CHECK:         firrtl.instance foo @ModuleNameConflict_A()
 // CHECK-NEXT:    firrtl.instance {{[_a-zA-Z0-9]+}} {lowerToBind,
 // CHECK-SAME:      @[[groupModule]](
+
+// -----
+// Layerblock lowering must allow a value to be captured twice.
+// https://github.com/llvm/circt/issues/6694
+
+firrtl.circuit "CaptureHardwareMultipleTimes" {
+  firrtl.layer @A bind {
+    firrtl.layer @B bind {}
+  }
+
+  // CHECK: firrtl.module private @[[A:.+]](in %[[p:.+]]: !firrtl.uint<1>) {
+  // CHECK:   %0 = firrtl.add %[[p]], %[[p]] : (!firrtl.uint<1>, !firrtl.uint<1>) -> !firrtl.uint<2>
+  // CHECK: }
+  // CHECK: firrtl.module @CaptureSrcTwice() {
+  // CHECK:   %c0_ui1 = firrtl.constant 0 : !firrtl.uint<1>
+  // CHECK:   %[[p:.+]] = firrtl.instance {{.+}} @[[A]]
+  // CHECK:   firrtl.strictconnect %[[p]], %c0_ui1 : !firrtl.uint<1>
+  // CHECK: }
+  firrtl.module @CaptureSrcTwice() {
+    %c0_ui1 = firrtl.constant 0 : !firrtl.uint<1>
+    firrtl.layerblock @A {
+      %0 = firrtl.add %c0_ui1, %c0_ui1 : (!firrtl.uint<1>, !firrtl.uint<1>) -> !firrtl.uint<2>
+    }
+  }
+
+  // CHECK: firrtl.module private @[[A:.+]](out %[[dst:.+]]: !firrtl.probe<uint<1>>, in %[[src:.+]]: !firrtl.uint<1>)
+  // CHECK:   %0 = firrtl.ref.send %[[src]] : !firrtl.uint<1>
+  // CHECK:   %c0_ui1 = firrtl.constant 0 : !firrtl.uint<1>
+  // CHECK:   %1 = firrtl.ref.send %c0_ui1 : !firrtl.uint<1>
+  // CHECK:   firrtl.ref.define %[[dst]], %1 : !firrtl.probe<uint<1>>
+  // CHECK:   %2 = firrtl.ref.resolve %0 : !firrtl.probe<uint<1>>
+  // CHECK: }
+  // CHECK: firrtl.module @CaptureAsDstThenSrc() {
+  // CHECK:   %w = firrtl.wire : !firrtl.probe<uint<1>>
+  // CHECK:   %[[out:.+]], %[[in:.+]] = firrtl.instance {{.+}} @[[A]](out {{.+}}: !firrtl.probe<uint<1>>, in {{.+}}: !firrtl.uint<1>)
+  // CHECK:   %0 = firrtl.ref.resolve %w : !firrtl.probe<uint<1>>
+  // CHECK:   firrtl.strictconnect %[[in]], %0 : !firrtl.uint<1>
+  // CHECK:   firrtl.ref.define %w, %[[out]] : !firrtl.probe<uint<1>>
+  // CHECK: }
+  firrtl.module @CaptureAsDstThenSrc() {
+    %w = firrtl.wire : !firrtl.probe<uint<1>, @A>
+    firrtl.layerblock @A {
+      // capture first as a sink.
+      %c0_ui1 = firrtl.constant 0 : !firrtl.uint<1>
+      %0 = firrtl.ref.send %c0_ui1 : !firrtl.uint<1>
+      %1 = firrtl.ref.cast %0 : (!firrtl.probe<uint<1>>) -> !firrtl.probe<uint<1>, @A>
```

Comment:
This could be simplified to a wire of the appropriate ref type.

Ditto in the dual of this test below.